client/client-multi/client-cpp/kaa/event/registration/EndpointRegistrationManager.hpp

```diff
@@ -83,52 +81,40 @@ class EndpointRegistrationManager : public IEndpointRegistrationManager
     virtual void onCurrentEndpointAttach(const UserAttachNotification& response);
     virtual void onCurrentEndpointDetach(const UserDetachNotification& response);
 
-    virtual bool isCurrentEndpointAttached() {
-        return status_->getEndpointAttachStatus();
-    }
-
-    virtual void setTransport(UserTransport * transport) {
-        userTransport_ = transport;
-        if (userTransport_ != nullptr && (userAttachRequest_.get() != nullptr || !attachingEndpoints_.empty() || !detachingEndpoints_.empty())) {
-            userTransport_->sync();
-        }
-    }
-
-    virtual void setAttachStatusListener(IEndpointAttachStatusListener* listener) {
-        if (listener != nullptr) {
-            attachStatusListener_ = listener;
-        }
-    }
+    void setTransport(UserTransport * transport);
 
 private:
     void onEndpointAccessTokenChanged(const std::string& old);
 
-private:
-    struct EndpointOperationInfo {
-        std::string endpointData_;/* endpoint's token or hash */
-        IEndpointAttachStatusListener* listener_;
-    };
+    void doSync();
 
+private:
+#ifdef KAA_THREADSAFE
+    typedef std::atomic_int_fast32_t RequestId;
+#else
+    typedef std::int32_t RequestId;
+#endif
 
 private:
-    IKaaClientStateStoragePtr     status_;
+    UserTransport*            userTransport_;
+    IKaaClientStateStoragePtr status_;
 
-    std::string endpointAccessToken_;
-    std::string endpointKeyHash_;
+    UserAttachRequestPtr      userAttachRequest_;
+    IUserAttachCallback*      userAttachResponseListener_;
+    IAttachStatusListener*    attachStatusListener_;
 
-    UserAttachRequestPtr userAttachRequest_;
+    RequestId attachRequestId_;
+    RequestId detachRequestId_;
 
-    std::map<std::int32_t/*requestId*/, EndpointOperationInfo>  attachingEndpoints_;
-    std::map<std::int32_t/*requestId*/, EndpointOperationInfo>  detachingEndpoints_;
-    std::map<std::string/*epToken*/, std::string/*epHash*/>    attachedEndpoints_;
-    KAA_R_MUTEX_DECLARE(endpointsGuard_);
+    KAA_MUTEX_DECLARE(attachEndpointGuard_);
+    KAA_MUTEX_DECLARE(detachEndpointGuard_);
 
-    UserTransport *                                            userTransport_;
+    std::unordered_map<std::int32_t, std::string/* endpoint access token */> attachEndpointRequests_;
+    std::unordered_map<std::int32_t, std::string/* endpoint key hash */>     detachEndpointRequests_;
 
-    KaaObservable<void (const AttachedEndpoints&), IAttachedEndpointListListener *> attachedEPListListeners_;
-    KAA_R_MUTEX_DECLARE(listenerGuard_);
+    std::unordered_map<std::int32_t, IAttachEndpointCallback*> attachEndpointListeners;
```

Comment:
What if the user will add a new Endpoint Listener during callback processing? The deadlock will occur. Why don't you use KaaObservable<void (const AttachedEndpoints&), std::int32_t> instead?